contracts/RewardManager.sol

```diff
@@ -0,0 +1,273 @@
+     * @param _burner Burner for collecting preMaturePenalty
+     * @dev deployer of contract on constructor is set as owner
+     */
+    constructor (
+        IERC20 _cnt,
+        uint256 _startDistribution,
+        uint256 _endDistribution,
+        uint256 _upfrontUnlock,
+        uint256 _preMaturePenalty,
+        uint256 _bonusPercentage,
+        address _burner) 
+        checkPercentages(_upfrontUnlock, _preMaturePenalty)
+        checkTime(_startDistribution, _endDistribution)
+
+    {
+        cnt = _cnt;
+        startDistribution = _startDistribution;
+        endDistribution = _endDistribution;
+        upfrontUnlock = _upfrontUnlock;
+        preMaturePenalty = _preMaturePenalty;
+        bonusPercentage = _bonusPercentage;
+        l2Burner = _burner;
+    }
+        
+    function _getNow() internal view returns (uint256) {
+        return block.timestamp;
+    }
+
+    function updatePreMaturePenalty(uint256 _newpreMaturePenalty) external 
+    checkPercentages(upfrontUnlock, _newpreMaturePenalty) 
+    onlyOwner
+    {
+        preMaturePenalty = _newpreMaturePenalty;
+    }
+
+    function updateBonusPercentage(uint256 _newBonusPercentage) external 
+    onlyOwner
+    {
+        bonusPercentage = _newBonusPercentage;
+    }
+    
+    function updateDistributionTime(uint256 _updatedStartTime, uint256 _updatedEndTime) external 
+    checkTime(_updatedStartTime, _updatedEndTime)
+    onlyOwner
+    {
+        require(startDistribution > _getNow(), "Vesting already started can't update now");
+        startDistribution = _updatedStartTime;
+        endDistribution = _updatedEndTime;
+    }
+    
+    function updateUpfrontUnlock(uint256 _newUpfrontUnlock) external 
+    checkPercentages(_newUpfrontUnlock, preMaturePenalty) 
+    onlyOwner
+    {
+        upfrontUnlock = _newUpfrontUnlock;
+    }
+
+    function updateWhitelistAddress(address _excludeAddress, bool status) external onlyOwner{
+        excludedAddresses[_excludeAddress] = status;
+    }
+
+    function updateRewardDistributor(address _distributor, bool status) external onlyOwner{
+        rewardDistributor[_distributor] = status;
+    }
+        
+    function handleRewardsForUser(
+        address user,
+        uint256 rewardAmount,
+        uint256 timestamp,
+        uint256 pid,
+        uint256 rewardDebt
+    ) external {
+        require(rewardDistributor[msg.sender],"Not a valid RewardDistributor");
+        if(rewardAmount > 0){
+            if(excludedAddresses[user]){
+                cnt.safeTransfer(user, rewardAmount);
+            }
+            else{
+                uint256 upfrontAmount = rewardAmount.mul(upfrontUnlock).div(1000);
+                cnt.safeTransfer(user, upfrontAmount);
+                vest(user, rewardAmount.sub(upfrontAmount));
+            }
+        }
+    }
+    
+    function vest(address _user, uint256 _amount) internal {
+        require(_getNow() < startDistribution, "Cannot vest");
```

Comment:
Revert Message: Cannot vest in distribution phase

Reply:
updated